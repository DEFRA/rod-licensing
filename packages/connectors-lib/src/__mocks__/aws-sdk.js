import { configureAwsSdkMock } from './aws-mock-helper.js'
const AwsSdk = configureAwsSdkMock()
export default AwsSdk
