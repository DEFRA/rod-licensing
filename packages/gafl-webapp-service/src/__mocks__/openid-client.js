export const mockSettings = {
  ENV: {
    OIDC_ENDPOINT: 'https://oauth-endpoint/',
    OIDC_REDIRECT_HOST: 'https://localhost:3143',
    OIDC_CLIENT_ID: 'client-id',
    OIDC_CLIENT_SECRET: 'client-secret',
    OIDC_SESSION_COOKIE_NAME: 'oidc-cookie-name',
    OIDC_SESSION_COOKIE_PASSWORD: 'open-sesame',
    OIDC_REQUIRE_DYNAMICS_ROLE: 'Telesales-role'
  },
  TEST_AUTH_ENDPOINT: 'https://oauth-endpoint/token-endpoint',
  TEST_AUTH_ISSUER: 'https://oauth-endpoint/issuer',
  TEST_NONCE: 'test_nonce',
  TEST_STATE: 'test_state',
  TEST_OID: 'test-azure-object-id',
  TEST_NAME: 'user fullname',
  TEST_EMAIL: 'user@example.com',
  TEST_EXP: 9999999999,
  TEST_POST_AUTH_REDIRECT: '/test/redirect'
}

const OpenIdClient = jest.genMockFromModule('openid-client')
OpenIdClient.Issuer = {
  discover: jest.fn(async () => ({
    issuer: mockSettings.TEST_AUTH_ISSUER,
    metadata: {},
    Client: mockOidcClientConstructor
  }))
}
OpenIdClient.generators.state = jest.fn(() => mockSettings.TEST_STATE)
OpenIdClient.generators.nonce = jest.fn(() => mockSettings.TEST_NONCE)

export const setupEnvironment = () => {
  Object.assign(process.env, mockSettings.ENV)
}

export const mockOidcClient = {
  authorizationUrl: jest.fn(() => mockSettings.TEST_AUTH_ENDPOINT),
  callback: jest.fn(() => ({
    claims: jest.fn(() => ({
      oid: mockSettings.TEST_OID,
      name: mockSettings.TEST_NAME,
      email: mockSettings.TEST_EMAIL,
      exp: mockSettings.TEST_EXP
    }))
  }))
}

export const mockOidcClientConstructor = jest.fn(() => mockOidcClient)

export const mockCache = {
  get: jest.fn(async () => {
    return {
      nonce: mockSettings.TEST_NONCE,
      state: mockSettings.TEST_STATE,
      postAuthRedirect: mockSettings.TEST_POST_AUTH_REDIRECT
    }
  }),
  set: jest.fn(),
  clear: jest.fn()
}
export const mockServer = {
  cache: jest.fn(() => mockCache),
  auth: {
    strategy: jest.fn(),
    default: jest.fn()
  }
}

export default OpenIdClient
