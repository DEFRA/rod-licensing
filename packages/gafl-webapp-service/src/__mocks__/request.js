const createCache = (cache = {}) => ({
  helpers: {
    status: {
      getCurrentPermission: () => cache.status || {},
      set: cache.status?.set || (() => {})
    },
    transaction: {
      getCurrentPermission: () => cache.transaction?.permissions[0] || {},
      get: () => cache.transaction || { permissions: [] }
    }
  }
})

export const createMockRequest = (opts = {}) => ({
  cache: () => createCache(opts.cache),
  payload: opts.payload || {},
  query: opts.query || {}
})

export const createMockRequestToolkit = () => ({
  redirect: jest.fn()
})
