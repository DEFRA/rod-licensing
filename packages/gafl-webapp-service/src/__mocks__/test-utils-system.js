import { createServer, init, server } from '../server.js'
import { SESSION_COOKIE_NAME_DEFAULT, CSRF_TOKEN_COOKIE_NAME_DEFAULT } from '../constants.js'
import { TEST_TRANSACTION, TEST_STATUS, GET_PRICING_TYPES, GET_PRICING_LENGTHS, LICENCE_LENGTH, LICENCE_TYPE } from '../uri.js'

import CatboxMemory from '@hapi/catbox-memory'
import { salesApi } from '@defra-fish/connectors-lib'
import mockPermits from './data/permits'
import mockPermitsConcessions from './data/permit-concessions'
import mockConcessions from './data/concessions'
import mockDefraCountries from './data/defra-country'
import { pricingDetail } from '../processors/pricing-summary'

process.env.SESSION_COOKIE_PASSWORD = 'A'.repeat(32)

createServer({
  cache: [
    {
      provider: {
        constructor: CatboxMemory
      }
    }
  ]
})

const start = async done => {
  server.events.on('start', () => {
    done()
  })

  // Add a helper route to expose the transaction cache
  server.route({
    method: 'GET',
    path: TEST_TRANSACTION.uri,
    handler: async request => request.cache().helpers.transaction.get()
  })

  server.route({
    method: 'GET',
    path: TEST_STATUS.uri,
    handler: async request => request.cache().helpers.status.get()
  })

  server.route({
    method: 'GET',
    path: GET_PRICING_TYPES.uri,
    handler: async request => pricingDetail(LICENCE_TYPE.page, await request.cache().helpers.transaction.getCurrentPermission())
  })

  server.route({
    method: 'GET',
    path: GET_PRICING_LENGTHS.uri,
    handler: async request => pricingDetail(LICENCE_LENGTH.page, await request.cache().helpers.transaction.getCurrentPermission())
  })

  // clear cache
  server.route({
    method: 'GET',
    path: '/buy/clear-cache',
    handler: async request => {
      await request.cache().clear()
      return 'ok'
    }
  })

  await init()
}

const stop = done => {
  server.events.on('stop', () => {
    done()
  })
  server.stop()
}

const getCookies = response => {
  const cookies = {}
  response.headers['set-cookie'] &&
    response.headers['set-cookie'].forEach(cookie => {
      const parts = cookie.split(';')[0].match(/(.*?)=(.*)$/)
      cookies[parts[1].trim()] = parts[2].trim()
    })
  return cookies
}

const initialize = async done => {
  const data = await server.inject({
    method: 'GET',
    url: '/buy'
  })

  global.cookies = getCookies(data)

  done()
}

const injectWithCookies = async (method, url, payload = {}) => {
  if (method === 'POST') {
    Object.assign(payload, { [CSRF_TOKEN_COOKIE_NAME_DEFAULT]: global.cookies[CSRF_TOKEN_COOKIE_NAME_DEFAULT] })
  }
  return server.inject({
    method,
    url,
    payload,
    headers: {
      cookie: `${SESSION_COOKIE_NAME_DEFAULT}=${global.cookies.sid}; ${CSRF_TOKEN_COOKIE_NAME_DEFAULT}=${global.cookies[CSRF_TOKEN_COOKIE_NAME_DEFAULT]}`
    }
  })
}

const injectWithoutSessionCookie = async (method, url, payload = {}) => {
  Object.assign(payload, { [CSRF_TOKEN_COOKIE_NAME_DEFAULT]: global.cookies[CSRF_TOKEN_COOKIE_NAME_DEFAULT] })
  return server.inject({
    method,
    url,
    payload,
    headers: { cookie: `${CSRF_TOKEN_COOKIE_NAME_DEFAULT}=${global.cookies[CSRF_TOKEN_COOKIE_NAME_DEFAULT]}` }
  })
}

const mockSalesApi = () => {
  salesApi.permits.getAll = jest.fn(async () => new Promise(resolve => resolve(mockPermits)))
  salesApi.permitConcessions.getAll = jest.fn(async () => new Promise(resolve => resolve(mockPermitsConcessions)))
  salesApi.concessions.getAll = jest.fn(async () => new Promise(resolve => resolve(mockConcessions)))
  salesApi.countries.getAll = jest.fn(async () => new Promise(resolve => resolve(mockDefraCountries)))
}

export { start, stop, server, getCookies, initialize, injectWithCookies, injectWithoutSessionCookie, mockSalesApi }
