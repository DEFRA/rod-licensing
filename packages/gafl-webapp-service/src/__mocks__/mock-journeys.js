import { injectWithCookies, mockSalesApi } from './test-utils-system.js'
import { dobHelper, ADULT_TODAY, JUNIOR_TODAY, SENIOR_TODAY } from './test-utils-business-rules.js'

import {
  ADDRESS_ENTRY,
  CONTACT,
  CONTACT_SUMMARY,
  CONTROLLER,
  DATE_OF_BIRTH,
  LICENCE_LENGTH,
  LICENCE_SUMMARY,
  LICENCE_TO_START,
  LICENCE_TYPE,
  NAME,
  NEW_TRANSACTION,
  NEWSLETTER,
  TERMS_AND_CONDITIONS,
  DISABILITY_CONCESSION
} from '../uri.js'
import { licenceToStart } from '../pages/licence-details/licence-to-start/update-transaction'
import { licenseTypes } from '../pages/licence-details/licence-type/route'

import { disabilityConcessionTypes } from '../pages/concessions/disability/update-transaction.js'

const goodAddress = {
  premises: '14 HOWECROFT COURT',
  street: 'EASTMEAD LANE',
  locality: '',
  town: 'BRISTOL',
  postcode: 'BS9 1HJ',
  'country-code': 'GB-ENG'
}

mockSalesApi()

export const ADULT_FULL_1_DAY_LICENCE = {
  transactionResponse: {
    id: 'ad28f2b3-2ab3-4d45-abc7-ae62c4da7944',
    expires: 1588342308,
    dataSource: 'Web Sales',
    permissions: [
      {
        permitId: 'a51b34a0-0c66-e611-80dc-c4346bad0190',
        licensee: {
          birthDate: '1970-01-01',
          firstName: 'Graham',
          lastName: 'Willis',
          premises: '14 Howecroft Court',
          street: 'Eastmead Lane',
          town: 'Bristol',
          postcode: 'BS9 1HJ',
          country: 'England',
          preferredMethodOfNewsletter: 'Prefer not to be contacted',
          preferredMethodOfConfirmation: 'Email',
          preferredMethodOfReminder: 'Email',
          email: 'new3@example.com',
          postalFulfilment: false
        },
        issueDate: '2020-04-29T14:11:47.855Z',
        startDate: '2020-04-28T23:00:00.000Z',
        referenceNumber: '01300420-1WS1FGW-B2F11U',
        endDate: '2020-04-29T23:00:00.000Z'
      }
    ],
    cost: 12
  },
  setup: async () => {
    // Initialize
    await injectWithCookies('GET', NEW_TRANSACTION.uri)
    await injectWithCookies('GET', CONTROLLER.uri)

    // Set up the licence details
    await injectWithCookies('POST', DATE_OF_BIRTH.uri, dobHelper(ADULT_TODAY))
    await injectWithCookies('POST', LICENCE_TO_START.uri, { 'licence-to-start': licenceToStart.AFTER_PAYMENT })
    await injectWithCookies('POST', LICENCE_TYPE.uri, { 'licence-type': licenseTypes.troutAndCoarse2Rod })
    await injectWithCookies('POST', LICENCE_LENGTH.uri, { 'licence-length': '1D' })
    await injectWithCookies('GET', LICENCE_SUMMARY.uri)

    // Set up the contact details
    await injectWithCookies('POST', NAME.uri, { 'last-name': 'Graham', 'first-name': 'Willis' })
    await injectWithCookies('POST', ADDRESS_ENTRY.uri, goodAddress)
    await injectWithCookies('POST', CONTACT.uri, { 'how-contacted': 'email', email: 'new3@example.com' })
    await injectWithCookies('POST', NEWSLETTER.uri, { newsletter: 'no' })
    await injectWithCookies('GET', CONTACT_SUMMARY.uri)

    // Complete the pre-payment journey
    await injectWithCookies('POST', CONTACT_SUMMARY.uri)
    await injectWithCookies('POST', TERMS_AND_CONDITIONS.uri, { agree: 'yes' })
  }
}

export const ADULT_DISABLED_12_MONTH_LICENCE = {
  transactionResponse: {
    id: '50c26295-013f-48c0-b354-f22544254fab',
    expires: 1588408572,
    dataSource: 'Web Sales',
    permissions: [
      {
        permitId: 'db1b34a0-0c66-e611-80dc-c4346bad0190',
        licensee: {
          birthDate: '1970-01-01',
          firstName: 'Graham',
          lastName: 'Willis',
          premises: '14 Howecroft Court',
          street: 'Eastmead Lane',
          town: 'Bristol',
          postcode: 'BS9 1HJ',
          country: 'England',
          preferredMethodOfNewsletter: 'Prefer not to be contacted',
          preferredMethodOfConfirmation: 'Letter',
          preferredMethodOfReminder: 'Letter',
          postalFulfilment: false
        },
        issueDate: '2020-04-30T08:36:12.195Z',
        startDate: '2020-04-29T23:00:00.000Z',
        concessions: [
          {
            id: 'd1ece997-ef65-e611-80dc-c4346bad4004',
            proof: {
              type: 'National Insurance Number',
              referenceNumber: 'QQ 12 34 56 C'
            }
          }
        ],
        referenceNumber: '01300421-1WS3FGW-YK1TTK',
        endDate: '2021-04-29T23:00:00.000Z'
      }
    ],
    cost: 82
  },
  setup: async () => {
    // Initialize
    await injectWithCookies('GET', NEW_TRANSACTION.uri)
    await injectWithCookies('GET', CONTROLLER.uri)

    // Set up the licence details
    await injectWithCookies('POST', DATE_OF_BIRTH.uri, dobHelper(ADULT_TODAY))
    await injectWithCookies('POST', LICENCE_TO_START.uri, { 'licence-to-start': licenceToStart.AFTER_PAYMENT })
    await injectWithCookies('POST', DISABILITY_CONCESSION.uri, {
      'disability-concession': disabilityConcessionTypes.pipDla,
      'ni-number': 'NH 34 67 44 A'
    })
    await injectWithCookies('POST', LICENCE_TYPE.uri, { 'licence-type': licenseTypes.troutAndCoarse2Rod })
    await injectWithCookies('POST', LICENCE_LENGTH.uri, { 'licence-length': '12M' })
    await injectWithCookies('GET', LICENCE_SUMMARY.uri)

    // Set up the contact details
    await injectWithCookies('POST', NAME.uri, { 'last-name': 'Graham', 'first-name': 'Willis' })
    await injectWithCookies('POST', ADDRESS_ENTRY.uri, goodAddress)
    await injectWithCookies('POST', CONTACT.uri, { 'how-contacted': 'email', email: 'new3@example.com' })
    await injectWithCookies('POST', NEWSLETTER.uri, { newsletter: 'no' })
    await injectWithCookies('GET', CONTACT_SUMMARY.uri)

    // Complete the pre-payment journey
    await injectWithCookies('POST', CONTACT_SUMMARY.uri)
    await injectWithCookies('POST', TERMS_AND_CONDITIONS.uri, { agree: 'yes' })
  }
}

export const SENIOR_12_MONTH_LICENCE = {
  transactionResponse: {
    id: 'cb913e36-a3ab-45a7-af6c-40af8befd931',
    expires: 1588411018,
    dataSource: 'Web Sales',
    permissions: [
      {
        permitId: 'dd1b34a0-0c66-e611-80dc-c4346bad0190',
        licensee: {
          birthDate: '1955-04-30',
          firstName: 'Graham',
          lastName: 'Willis',
          premises: '14 Howecroft Court',
          street: 'Eastmead Lane',
          town: 'Bristol',
          postcode: 'BS9 1HJ',
          country: 'England',
          preferredMethodOfNewsletter: 'Prefer not to be contacted',
          preferredMethodOfConfirmation: 'Letter',
          preferredMethodOfReminder: 'Letter',
          postalFulfilment: false
        },
        issueDate: '2020-04-30T09:16:58.635Z',
        startDate: '2020-04-29T23:00:00.000Z',
        concessions: [
          {
            id: 'd0ece997-ef65-e611-80dc-c4346bad4004',
            proof: {
              type: 'No proof'
            }
          }
        ],
        referenceNumber: '01300420-1WS0SGW-78CGFU',
        endDate: '2020-04-29T23:00:00.000Z'
      }
    ],
    cost: 54
  },
  setup: async () => {
    // Initialize
    await injectWithCookies('GET', NEW_TRANSACTION.uri)
    await injectWithCookies('GET', CONTROLLER.uri)

    // Set up the licence details
    await injectWithCookies('POST', DATE_OF_BIRTH.uri, dobHelper(SENIOR_TODAY))
    await injectWithCookies('POST', LICENCE_TO_START.uri, { 'licence-to-start': licenceToStart.AFTER_PAYMENT })
    await injectWithCookies('POST', LICENCE_TYPE.uri, { 'licence-type': licenseTypes.troutAndCoarse2Rod })
    await injectWithCookies('POST', LICENCE_LENGTH.uri, { 'licence-length': '12M' })
    await injectWithCookies('GET', LICENCE_SUMMARY.uri)

    // Set up the contact details
    await injectWithCookies('POST', NAME.uri, { 'last-name': 'Graham', 'first-name': 'Willis' })
    await injectWithCookies('POST', ADDRESS_ENTRY.uri, goodAddress)
    await injectWithCookies('POST', CONTACT.uri, { 'how-contacted': 'email', email: 'new3@example.com' })
    await injectWithCookies('POST', NEWSLETTER.uri, { newsletter: 'no' })
    await injectWithCookies('GET', CONTACT_SUMMARY.uri)

    // Complete the pre-payment journey
    await injectWithCookies('POST', CONTACT_SUMMARY.uri)
    await injectWithCookies('POST', TERMS_AND_CONDITIONS.uri, { agree: 'yes' })
  }
}

export const JUNIOR_LICENCE = {
  transactionResponse: {
    id: '8793ff10-6372-4e9c-b4f8-d0cde0ed2277',
    expires: 1588412128,
    dataSource: 'Web Sales',
    permissions: [
      {
        permitId: 'd91b34a0-0c66-e611-80dc-c4346bad0190',
        licensee: {
          birthDate: '2006-01-01',
          firstName: 'Graham',
          lastName: 'Willis',
          premises: '14 Howecroft Court',
          street: 'Eastmead Lane',
          town: 'Bristol',
          postcode: 'BS9 1HJ',
          country: 'England',
          preferredMethodOfNewsletter: 'Prefer not to be contacted',
          preferredMethodOfConfirmation: 'Prefer not to be contacted',
          preferredMethodOfReminder: 'Prefer not to be contacted',
          postalFulfilment: false
        },
        issueDate: '2020-04-30T09:35:28.014Z',
        startDate: '2020-04-29T23:00:00.000Z',
        concessions: [
          {
            id: '3230c68f-ef65-e611-80dc-c4346bad4004',
            proof: {
              type: 'No proof'
            }
          }
        ],
        referenceNumber: '01300420-1WS0JGW-FLNA84',
        endDate: '2020-04-29T23:00:00.000Z'
      }
    ],
    cost: 0
  },
  setup: async () => {
    // Initialize
    await injectWithCookies('GET', NEW_TRANSACTION.uri)
    await injectWithCookies('GET', CONTROLLER.uri)

    // Set up the licence details
    await injectWithCookies('POST', DATE_OF_BIRTH.uri, dobHelper(JUNIOR_TODAY))
    await injectWithCookies('POST', LICENCE_TO_START.uri, { 'licence-to-start': licenceToStart.AFTER_PAYMENT })
    await injectWithCookies('POST', LICENCE_TYPE.uri, { 'licence-type': licenseTypes.troutAndCoarse2Rod })
    await injectWithCookies('GET', LICENCE_SUMMARY.uri)

    // Set up the contact details
    await injectWithCookies('POST', NAME.uri, { 'last-name': 'Graham', 'first-name': 'Willis' })
    await injectWithCookies('POST', ADDRESS_ENTRY.uri, goodAddress)
    await injectWithCookies('POST', CONTACT.uri, { 'how-contacted': 'email', email: 'new3@example.com' })
    await injectWithCookies('POST', NEWSLETTER.uri, { newsletter: 'no' })
    await injectWithCookies('GET', CONTACT_SUMMARY.uri)

    // Complete the pre-payment journey
    await injectWithCookies('POST', CONTACT_SUMMARY.uri)
    await injectWithCookies('POST', TERMS_AND_CONDITIONS.uri, { agree: 'yes' })
  }
}

export const JUNIOR_DISABLED_LICENCE = {
  transactionResponse: {
    id: '8793ff10-6372-4e9c-b4f8-d0cde0ed2277',
    expires: 1588412128,
    dataSource: 'Web Sales',
    permissions: [
      {
        permitId: 'd91b34a0-0c66-e611-80dc-c4346bad0190',
        licensee: {
          birthDate: '2006-01-01',
          firstName: 'Graham',
          lastName: 'Willis',
          premises: '14 Howecroft Court',
          street: 'Eastmead Lane',
          town: 'Bristol',
          postcode: 'BS9 1HJ',
          country: 'England',
          preferredMethodOfNewsletter: 'Prefer not to be contacted',
          preferredMethodOfConfirmation: 'Prefer not to be contacted',
          preferredMethodOfReminder: 'Prefer not to be contacted',
          postalFulfilment: false
        },
        issueDate: '2020-04-30T09:35:28.014Z',
        startDate: '2020-04-29T23:00:00.000Z',
        concessions: [
          {
            id: 'd1ece997-ef65-e611-80dc-c4346bad4004',
            proof: {
              type: 'National Insurance Number',
              referenceNumber: 'NH 34 67 44 A'
            }
          }
        ],
        referenceNumber: '01300420-1WS0JGW-FLNA84',
        endDate: '2020-04-29T23:00:00.000Z'
      }
    ],
    cost: 0
  },
  setup: async () => {
    // Initialize
    await injectWithCookies('GET', NEW_TRANSACTION.uri)
    await injectWithCookies('GET', CONTROLLER.uri)

    // Set up the licence details
    await injectWithCookies('POST', DATE_OF_BIRTH.uri, dobHelper(JUNIOR_TODAY))
    await injectWithCookies('POST', LICENCE_TO_START.uri, { 'licence-to-start': licenceToStart.AFTER_PAYMENT })
    await injectWithCookies('POST', DISABILITY_CONCESSION.uri, {
      'disability-concession': disabilityConcessionTypes.pipDla,
      'ni-number': 'NH 34 67 44 A'
    })
    await injectWithCookies('POST', LICENCE_TYPE.uri, { 'licence-type': licenseTypes.troutAndCoarse2Rod })
    await injectWithCookies('GET', LICENCE_SUMMARY.uri)

    // Set up the contact details
    await injectWithCookies('POST', NAME.uri, { 'last-name': 'Graham', 'first-name': 'Willis' })
    await injectWithCookies('POST', ADDRESS_ENTRY.uri, goodAddress)
    await injectWithCookies('POST', CONTACT.uri, { 'how-contacted': 'email', email: 'new3@example.com' })
    await injectWithCookies('POST', NEWSLETTER.uri, { newsletter: 'no' })
    await injectWithCookies('GET', CONTACT_SUMMARY.uri)

    // Complete the pre-payment journey
    await injectWithCookies('POST', CONTACT_SUMMARY.uri)
    await injectWithCookies('POST', TERMS_AND_CONDITIONS.uri, { agree: 'yes' })
  }
}

export const MOCK_PAYMENT_RESPONSE = {
  state: { status: 'created', finished: false },
  payment_id: 'qdq15eu98cpk8bc14qs9ht0t3v',
  payment_provider: 'sandbox',
  created_date: '2020-05-05T07:30:56.214Z',
  _links: {
    next_url: {
      href: 'https://www.payments.service.gov.uk/secure/017f99a4-977d-40c2-8a2f-fb0f995a88f0',
      method: 'GET'
    },
    self: {
      href: 'https://publicapi.payments.service.gov.uk/v1/payments/qdq15eu98cpk8bc14qs9ht0t3v',
      method: 'GET'
    }
  }
}
