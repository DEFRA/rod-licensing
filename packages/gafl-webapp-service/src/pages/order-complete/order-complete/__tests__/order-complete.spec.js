import { salesApi } from '@defra-fish/connectors-lib'

import { initialize, injectWithCookies, start, stop, mockSalesApi } from '../../../../__mocks__/test-utils-system'
import { JUNIOR_LICENCE } from '../../../../__mocks__/mock-journeys.js'
import {
  AGREED,
  ORDER_COMPLETE,
  TERMS_AND_CONDITIONS,
  LICENCE_INFORMATION,
  COOKIES,
  ACCESSIBILITY_STATEMENT,
  PRIVACY_POLICY,
  REFUND_POLICY
} from '../../../../uri.js'

jest.mock('@defra-fish/connectors-lib')
mockSalesApi()

beforeAll(() => {
  process.env.ANALYTICS_PRIMARY_PROPERTY = 'UA-123456789-0'
  process.env.ANALYTICS_XGOV_PROPERTY = 'UA-987654321-0'
})
beforeAll(d => start(d))
beforeAll(d => initialize(d))
afterAll(d => stop(d))
afterAll(() => {
  delete process.env.ANALYTICS_PRIMARY_PROPERTY
  delete process.env.ANALYTICS_XGOV_PROPERTY
})

describe('The order completion handler', () => {
  it('throws a status 403 (forbidden) exception if the agreed flag is not set', async () => {
    const data = await injectWithCookies('GET', ORDER_COMPLETE.uri)
    expect(data.statusCode).toBe(403)
  })

  it('throws a status 403 (forbidden) exception if the posted flag is not set', async () => {
    await injectWithCookies('POST', TERMS_AND_CONDITIONS.uri, { agree: 'yes' })
    const data = await injectWithCookies('GET', ORDER_COMPLETE.uri)
    expect(data.statusCode).toBe(403)
  })

  it('throw a status 403 (forbidden) exception if the finalized flag is not set', async () => {
    await JUNIOR_LICENCE.setup()
    salesApi.createTransaction.mockResolvedValue(JUNIOR_LICENCE.transactionResponse)
    salesApi.finaliseTransaction.mockRejectedValue(new Error())

    await injectWithCookies('GET', AGREED.uri)
    const data = await injectWithCookies('GET', ORDER_COMPLETE.uri)
    expect(data.statusCode).toBe(403)
  })

  it('responds with the order completed page if the journey has finished', async () => {
    await JUNIOR_LICENCE.setup()
    salesApi.createTransaction.mockResolvedValue(JUNIOR_LICENCE.transactionResponse)
    salesApi.finaliseTransaction.mockResolvedValue(JUNIOR_LICENCE.transactionResponse)

    const data1 = await injectWithCookies('GET', AGREED.uri)
    expect(data1.statusCode).toBe(302)
    expect(data1.headers.location).toBe(ORDER_COMPLETE.uri)
    const data = await injectWithCookies('GET', ORDER_COMPLETE.uri)
    expect(data.statusCode).toBe(200)
  })

  it('sets the currentPage to order-complete in the cache', async () => {
    await JUNIOR_LICENCE.setup()
    salesApi.createTransaction.mockResolvedValue(JUNIOR_LICENCE.transactionResponse)
    salesApi.finaliseTransaction.mockResolvedValue(JUNIOR_LICENCE.transactionResponse)

    await injectWithCookies('GET', AGREED.uri)
    const data = await injectWithCookies('GET', ORDER_COMPLETE.uri)
    const permission = await data.request.cache().helpers.status.getCurrentPermission()
    expect(permission.currentPage).toBe(ORDER_COMPLETE.page)
  })

  it('responds with the licence information page when requested', async () => {
    await JUNIOR_LICENCE.setup()
    salesApi.createTransaction.mockResolvedValue(JUNIOR_LICENCE.transactionResponse)
    salesApi.finaliseTransaction.mockResolvedValue(JUNIOR_LICENCE.transactionResponse)

    await injectWithCookies('GET', AGREED.uri)
    await injectWithCookies('GET', ORDER_COMPLETE.uri)
    const data = await injectWithCookies('GET', LICENCE_INFORMATION.uri)
    expect(data.statusCode).toBe(200)
  })

  it.each([
    [COOKIES.page, COOKIES.uri],
    [ACCESSIBILITY_STATEMENT.page, ACCESSIBILITY_STATEMENT.uri],
    [PRIVACY_POLICY.page, PRIVACY_POLICY.uri],
    [REFUND_POLICY.page, REFUND_POLICY.uri]
  ])('succesfully navigates to %s when on the order complete page', async (page, uri) => {
    await JUNIOR_LICENCE.setup()
    salesApi.createTransaction.mockResolvedValue(JUNIOR_LICENCE.transactionResponse)
    salesApi.finaliseTransaction.mockResolvedValue(JUNIOR_LICENCE.transactionResponse)

    await injectWithCookies('GET', AGREED.uri)
    await injectWithCookies('GET', ORDER_COMPLETE.uri)

    const data = await injectWithCookies('GET', uri)
    expect(data.statusCode).toBe(200)
  })
})
