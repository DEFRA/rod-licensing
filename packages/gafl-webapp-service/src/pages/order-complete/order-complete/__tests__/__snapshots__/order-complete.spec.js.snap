// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The order completion handler 8 day licence sets as non_postal 1`] = `
Object {
  "futurePaymentsParagraphOne": "future payments digital one",
  "futurePaymentsParagraphTwo": "future payments digital two",
  "licenceDetailsDigitalParagraph": "fishing self digital confirmation licence details",
  "licenceDetailsParagraphTwo": "paragraph self licence details",
  "licenceTitle": "title self licence details",
  "title": "payment title 1",
  "whenFishingParagraphOne": "fishing self non postal digital when fishing",
  "whenFishingParagraphOneLink": "fishing self link",
  "whenFishingParagraphTwo": "fishing self non postal non digital when fishing two",
}
`;

exports[`The order completion handler 12 month digital licence for someone else 1`] = `
Object {
  "futurePaymentsParagraphOne": "future payments digital one",
  "futurePaymentsParagraphTwo": "future payments digital two",
  "licenceDetailsDigitalParagraph": "fishing bobo digital licence details",
  "licenceDetailsParagraphTwo": "paragraph bobo licence details",
  "licenceTitle": "title bobo licence details",
  "title": "payment title 1",
  "whenFishingParagraphOne": "fishing bobo non postal digital when fishing",
  "whenFishingParagraphOneLink": "fishing bobo link",
  "whenFishingParagraphTwo": "fishing bobo non postal non digital when fishing two",
}
`;

exports[`The order completion handler 12 month digital licence for you 1`] = `
Object {
  "futurePaymentsParagraphOne": "future payments digital one",
  "futurePaymentsParagraphTwo": "future payments digital two",
  "licenceDetailsDigitalParagraph": "fishing self digital licence details",
  "licenceDetailsParagraphTwo": "paragraph self licence details",
  "licenceTitle": "title self licence details",
  "title": "payment title 1",
  "whenFishingParagraphOne": "fishing self non postal digital when fishing",
  "whenFishingParagraphOneLink": "fishing self link",
  "whenFishingParagraphTwo": "fishing self non postal non digital when fishing two",
}
`;

exports[`The order completion handler 12 month postal licence for someone else with digital confirmation 1`] = `
Object {
  "futurePaymentsParagraphOne": "future payments digital one",
  "futurePaymentsParagraphTwo": "future payments digital two",
  "licenceDetailsDigitalParagraph": "fishing bobo digital confirmation licence details",
  "licenceDetailsParagraphTwo": "paragraph bobo licence details",
  "licenceTitle": "title bobo licence details",
  "title": "payment title 1",
  "whenFishingParagraphOne": "fishing bobo postal digital when fishing",
  "whenFishingParagraphOneLink": "fishing bobo link",
  "whenFishingParagraphTwo": "fishing bobo non postal non digital when fishing two",
}
`;

exports[`The order completion handler 12 month postal licence for someone else with none digital confirmation 1`] = `
Object {
  "futurePaymentsParagraphOne": "future payments digital one",
  "futurePaymentsParagraphTwo": "future payments digital two",
  "licenceDetailsDigitalParagraph": undefined,
  "licenceDetailsParagraphTwo": "paragraph bobo licence details",
  "licenceTitle": "title bobo licence details",
  "title": "payment title 1",
  "whenFishingParagraphOne": "fishing bobo non postal non digital when fishing",
  "whenFishingParagraphOneLink": "fishing bobo link",
  "whenFishingParagraphTwo": "fishing bobo non postal non digital when fishing two",
}
`;

exports[`The order completion handler 12 month postal licence for you with digital confirmation 1`] = `
Object {
  "futurePaymentsParagraphOne": "future payments digital one",
  "futurePaymentsParagraphTwo": "future payments digital two",
  "licenceDetailsDigitalParagraph": "fishing self digital confirmation licence details",
  "licenceDetailsParagraphTwo": "paragraph self licence details",
  "licenceTitle": "title self licence details",
  "title": "payment title 1",
  "whenFishingParagraphOne": "fishing self postal digital when fishing",
  "whenFishingParagraphOneLink": "fishing self link",
  "whenFishingParagraphTwo": "fishing self non postal non digital when fishing two",
}
`;

exports[`The order completion handler 12 month postal licence for you with none digital confirmation 1`] = `
Object {
  "futurePaymentsParagraphOne": "future payments digital one",
  "futurePaymentsParagraphTwo": "future payments digital two",
  "licenceDetailsDigitalParagraph": undefined,
  "licenceDetailsParagraphTwo": "paragraph self licence details",
  "licenceTitle": "title self licence details",
  "title": "payment title 1",
  "whenFishingParagraphOne": "fishing self non postal non digital when fishing",
  "whenFishingParagraphOneLink": "fishing self link",
  "whenFishingParagraphTwo": "fishing self non postal non digital when fishing two",
}
`;

exports[`The order completion handler 12 month recurring payment with digital reminder 1`] = `
Object {
  "futurePaymentsParagraphOne": "future payments digital one",
  "futurePaymentsParagraphTwo": "future payments digital two",
  "licenceDetailsDigitalParagraph": "fishing self digital licence details",
  "licenceDetailsParagraphTwo": "paragraph self licence details",
  "licenceTitle": "title self licence details",
  "title": "payment title 1",
  "whenFishingParagraphOne": "fishing self non postal digital when fishing",
  "whenFishingParagraphOneLink": "fishing self link",
  "whenFishingParagraphTwo": "fishing self non postal non digital when fishing two",
}
`;

exports[`The order completion handler 12 month recurring payment with postal reminder 1`] = `
Object {
  "futurePaymentsParagraphOne": "future payments postal one",
  "futurePaymentsParagraphTwo": "future payments postal two",
  "licenceDetailsDigitalParagraph": "fishing self digital licence details",
  "licenceDetailsParagraphTwo": "paragraph self licence details",
  "licenceTitle": "title self licence details",
  "title": "payment title 1",
  "whenFishingParagraphOne": "fishing self non postal digital when fishing",
  "whenFishingParagraphOneLink": "fishing self link",
  "whenFishingParagraphTwo": "fishing self non postal non digital when fishing two",
}
`;
